### PR TITLE
Update BrokerFailureDetector to use AdminClient

### PR DESCRIPTION
**Pull request by mimaison**
This removes the dependency on ZooKeeper of this component.

Without ZooKeeper, we can't get notified when a broker leaves the cluster. Instead we have to constantly query the cluster metadata and see which brokers are available.

### CODE REVIEW

**Review by efeg — COMMENTED** [review]
@mimaison Thanks for the PR!
Rather than removing the ZK-based broker failure detection logic, should we make the source of broker failure detection configurable? That way, if a user wants to use non-ZK-based broker failure detector, they can move to it.

**Comment by mimaison** [comment]
@efeg I've made the detection mechanism configurable. I experimented with having different classes but as both logic share a lot of code but don't have the same API, it did not help a lot.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManagerTest.java [hunk 1]
```diff
@@ -95,12 +95,14 @@ private static void startRunnableDetectors(ScheduledExecutorService mockDetector
                                              GoalViolationDetector mockGoalViolationDetector,
                                              MetricAnomalyDetector mockMetricAnomalyDetector,
                                              DiskFailureDetector mockDiskFailureDetector,
+                                             BrokerFailureDetector mockBrokerFailureDetector,
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManagerTest.java:98]: Should we update `BrokerFailureIntegrationTest` or add another integ. test for the adminClient-based broker failure detector?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnomalyDetectorConfig.java [hunk 2]
```diff
@@ -189,6 +189,24 @@
...
+  /**
+   * <code>broker.failure.detection.interval.ms</code>
+   */
+  public static final String BROKER_FAILURE_DETECTION_INTERVAL_MS_CONFIG = "broker.failure.detection.interval.ms";
+  public static final Long DEFAULT_BROKER_FAILURE_DETECTION_INTERVAL_MS = null;
+  public static final String BROKER_FAILURE_DETECTION_INTERVAL_MS_DOC = "The interval in millisecond that broker failure "
+          + "detector will run to detect broker failures. If this interval time is not specified, disk failure detector "
+          + "will run with interval specified in " + ANOMALY_DETECTION_INTERVAL_MS_CONFIG + ". This is only used when "
+          + KAFKA_BROKER_FAILURE_DETECTION_ENABLE_CONFIG + " is set to 'true'.";
+
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnomalyDetectorConfig.java:209]: Does the documentation or use of `broker.failure.detection.backoff.ms` change?

Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnomalyDetectorConfig.java:209]: 1. How do we handle the use of (1) `broker.failure.alert.threshold.ms` and (2) `broker.failure.self.healing.threshold.ms` with adminClient-based broker failure detector?
2. Do we need to update the JavaDoc of `SelfHealingNotifier`?

**Review by efeg — COMMENTED** [review]
Thanks for the update -- left some comment.

**Comment by efeg** [comment]
@mimaison Thanks for the update -- I see some build failure:
```
javadoc: error - An internal exception has occurred. 
	(com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.yammer.metrics.core.Gauge not found)
Please file a bug against the javadoc tool via the Java bug reporting page
(http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com)
for duplicates. Include error messages and the following diagnostic in your report. Thank you.
com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.yammer.metrics.core.Gauge not found
1 error
71 warnings

> Task :cruise-control:javadoc FAILED
```

**Comment by mimaison** [comment]
> @mimaison Thanks for the update -- I see some build failure:
> 
> ```
> javadoc: error - An internal exception has occurred. 
> 	(com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.yammer.metrics.core.Gauge not found)
> Please file a bug against the javadoc tool via the Java bug reporting page
> (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com)
> for duplicates. Include error messages and the following diagnostic in your report. Thank you.
> com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.yammer.metrics.core.Gauge not found
> 1 error
> 71 warnings
> 
> > Task :cruise-control:javadoc FAILED
> ```

Not quite sure how my PR is triggering that, it doesn't include significant javadoc changes or anything related to `com.yammer.metrics.core.Gauge`. It's working fine on Java 17 but I'm able to reproduce on Java 11.

**Comment by mimaison** [comment]
@efeg I pushed an update the fix the failure. Can you take another look? Thanks

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManagerTest.java [hunk 3]
```diff
@@ -172,10 +174,11 @@ public void testDelayedCheck() throws InterruptedException {
             .andReturn(AnomalyNotificationResult.check(MOCK_DELAY_CHECK_MS));
     EasyMock.expect(mockAnomalyNotifier.selfHealingEnabledRatio()).andReturn(MOCK_SELF_HEALING_ENABLED_RATIO);
     Properties props = KafkaCruiseControlUnitTestUtils.getKafkaCruiseControlProperties();
+    props.setProperty(AnomalyDetectorConfig.KAFKA_BROKER_FAILURE_DETECTION_ENABLE_CONFIG, "true");
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManagerTest.java:177]: Given that we are mocking `ZKBrokerFailureDetector`, shouldn't this be `false`?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetectorTest.java [hunk 4]
```diff
@@ -197,4 +228,14 @@ private void restartDeadBroker(int index) {
   private MockTime getMockTime() {
     return new MockTime(0, 100L, TimeUnit.NANOSECONDS.convert(100L, TimeUnit.MILLISECONDS));
   }
+
+  private void detect(AbstractBrokerFailureDetector detector, Queue<Anomaly> anomalies, long start, int count) {
+    if (_useKafkaApi) {
+      detector.run();
+    } else {
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetectorTest.java:235]: In a real execution, isn't `run()` called regardless of whether `_useKafkaApi` is `true` or not? To better emulate the actual run, shouldn't we always call `detector.run();`?

**Review by efeg — COMMENTED** [review]
@mimaison Thanks for the update -- left some comments.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AbstractBrokerFailureDetector.java [hunk 5]
```diff
@@ -6,95 +6,69 @@
...
-    _kafkaZkClient.getChildren(BrokerIdsZNode.path());
-  }
-
-  private synchronized void detectBrokerFailures(Set<Integer> aliveBrokers, boolean skipReportingIfNotUpdated) {
+  synchronized void detectBrokerFailures(boolean skipReportingIfNotUpdated) {
+    try {
+      _aliveBrokers = aliveBrokers();
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      LOG.error("Failed retrieving alive brokers", e);
+      _aliveBrokers = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (mimaison) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AbstractBrokerFailureDetector.java]: Clearing `_aliveBrokers` basically marks all brokers as failed. I don't think it's the best approach considering a timeout is a lot more likely than all brokers failing at the same time.

I've updated the logic to instead simply log the failure. At the risk of detecting a real failure a bit later, it's a bit more defensive. It is also inline with the behavior of the other detectors.

**Comment by mimaison** [comment]
Thanks for the review @efeg ! I've pushed an update

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/KafkaBrokerFailureDetector.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  @Override
+  public void run() {
+    detectBrokerFailures(false);
+  }
+
+  Set<Integer> aliveBrokers() throws ExecutionException, InterruptedException, TimeoutException {
+    return _adminClient.describeCluster().nodes().get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+              .stream().map(Node::id).collect(toSet());
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/KafkaBrokerFailureDetector.java:43]: Nit: Bad indentation.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/ZKBrokerFailureDetector.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  private boolean _started = false;
+
+  public ZKBrokerFailureDetector(Queue<Anomaly> anomalies, KafkaCruiseControl kafkaCruiseControl) {
+    super(anomalies, kafkaCruiseControl);
+    KafkaCruiseControlConfig config = _kafkaCruiseControl.config();
+    String zkUrl = config.getString(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG);
+    boolean zkSecurityEnabled = config.getBoolean(ExecutorConfig.ZOOKEEPER_SECURITY_ENABLED_CONFIG);
+    ZKClientConfig zkClientConfig = ZKConfigUtils.zkClientConfigFromKafkaConfig(config);
+    _kafkaZkClient = KafkaCruiseControlUtils.createKafkaZkClient(zkUrl, ZK_BROKER_FAILURE_METRIC_GROUP, ZK_BROKER_FAILURE_METRIC_TYPE,
+            zkSecurityEnabled, zkClientConfig);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/ZKBrokerFailureDetector.java:47]: Nit: Bad indentation.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/ZKBrokerFailureDetector.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      }
+    }
+    KafkaCruiseControlUtils.closeKafkaZkClientWithTimeout(_kafkaZkClient);
+  }
+
+  @Override
+  Set<Integer> aliveBrokers() {
+    // We get the alive brokers from ZK directly.
+    return CollectionConverters.asJava(_kafkaZkClient.getAllBrokersInCluster())
+            .stream().map(Broker::id).collect(toSet());
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/ZKBrokerFailureDetector.java:94]: Nit: Bad indentation.

**Review by efeg — APPROVED** [review]
Thanks for all the updates -- with your PR, Cruise Control can now support management of Kafka clusters w/o `ZooKeeper`!

I left a few nits on formatting / indentation.

**Comment by mimaison** [comment]
Thanks @efeg for the review! I've fixed the formatting issues.

If I understand correctly this was the last component with a hard dependency on ZooKeeper. The next step is probably a final PR to make the zookeeper configurations optional if `kafka.broker.failure.detection.enable` is enabled.

It would also be nice to have some integration tests with Kafka running in KRaft mode.

**Pull request merged**